packages/babel-plugin-transform-regenerator/src/regenerator/emit.ts

```diff
@@ -56,1288 +30,1301 @@ const Ep = Emitter.prototype;
-      assert.ok(thisLocValue >= lastLocValue, "try entries out of order");
-      lastLocValue = thisLocValue;
+        const locs = [
+          tryEntry.firstLoc,
+          // The null here makes a hole in the array.
+          ce ? ce.firstLoc : null,
+        ];
 
-      const ce = tryEntry.catchEntry;
-      const fe = tryEntry.finallyEntry;
+        if (fe) {
+          locs[2] = fe.firstLoc;
+          locs[3] = fe.afterLoc;
+        }
 
-      const locs = [
-        tryEntry.firstLoc,
-        // The null here makes a hole in the array.
-        ce ? ce.firstLoc : null,
-      ];
+        return t.arrayExpression(locs.map(loc => loc && t.cloneNode(loc)));
+      }),
+    );
+  }
 
-      if (fe) {
-        locs[2] = fe.firstLoc;
-        locs[3] = fe.afterLoc;
-      }
+  // All side effects must be realized in order.
 
-      return t.arrayExpression(locs.map(loc => loc && t.clone(loc)));
-    }),
-  );
-};
+  // If any subexpression harbors a leap, all subexpressions must be
+  // neutered of side effects.
 
-// All side effects must be realized in order.
+  // No destructive modification of AST nodes.
+  explode(path: NodePath, ignoreResult?: boolean) {
+    const node = path.node;
+    const self = this;
 
-// If any subexpression harbors a leap, all subexpressions must be
-// neutered of side effects.
+    if (t.isDeclaration(node)) throw getDeclError(node);
 
-// No destructive modification of AST nodes.
+    if (path.isStatement()) return self.explodeStatement(path);
 
-Ep.explode = function (path: any, ignoreResult: any) {
-  const t = util.getTypes();
-  const node = path.node;
-  const self = this;
+    if (path.isExpression()) return self.explodeExpression(path, ignoreResult);
 
-  t.assertNode(node);
+    switch (node.type) {
+      case "VariableDeclarator":
+        throw getDeclError(node);
 
-  if (t.isDeclaration(node)) throw getDeclError(node);
+      // These node types should be handled by their parent nodes
+      // (ObjectExpression, SwitchStatement, and TryStatement, respectively).
+      case "ObjectProperty":
+      case "SwitchCase":
+      case "CatchClause":
+        throw new Error(
+          node.type + " nodes should be handled by their parents",
+        );
 
-  if (t.isStatement(node)) return self.explodeStatement(path);
+      default:
+        throw new Error("unknown Node of type " + JSON.stringify(node.type));
+    }
+  }
 
-  if (t.isExpression(node)) return self.explodeExpression(path, ignoreResult);
+  explodeStatement(path: NodePath<t.Statement>, labelId: t.Identifier = null) {
+    const stmt = path.node;
+    const self = this;
+    let before: t.NumericLiteral,
+      after: t.NumericLiteral,
+      head: t.NumericLiteral;
+
+    // Explode BlockStatement nodes even if they do not contain a yield,
+    // because we don't want or need the curly braces.
+    if (t.isBlockStatement(stmt)) {
+      (path.get("body") as NodePath<t.Statement>[]).forEach(function (path) {
```

Comment:
I believe if we replace `t.isBlockStatement(stmt)` with `path.isBlockStatement()`, then TS will correctly infer `path.get("body")`.